### DISCUSSION TITLE
Can rewrite and path segment be used together?

### DISCUSSION THREAD

**Discussion by p-chan** [vercel/community#111] [Help] [answered]
I'm trying to create a REST API with Serverless Functions.

I will be using a custom domain like `api.example.com`, I decided to rewrite `/api/*` to `/*`.

`vercel.json`

```json
{
  "rewrites": [
    {
      "source": "/favicon.ico",
      "destination": "/public/favicon.ico"
    },
    {
      "source": "/(.*)",
      "destination": "/api/$1"
    },
    {
      "source": "/(.*)",
      "destination": "/api/404.ts"
    }
  ]
}
```

Then, I can access `/api/hello.ts` with `/hello`.

However, when I combine rewrites with path segment, it does not work.

For example, I create `/api/say/[name].ts` and when I access `/say/example`, I expected it to return 200. However, the endpoint returned a 404.

Is there any way to use rewrites and path segments together?

**Comment by stevefrench39** [comment]
We ran into the same problem with an api versioning strategy.

```
"rewrites": [
    {
      "source": "/v/:version/:rest*",
      "destination": "/api/:rest*?version=:version"
    }
]
```

This rewrite works for simple pattern `/v/1/accounts`, but fails for the path segment pattern `/v/1/accounts/[accountId]` with an incoming request like `GET /v/1/accounts/123`. 

The non-rewritten path segment pattern `/api/accounts/[accountId]` works as expected with an incoming request like `GET /api/accounts/123`.

Incidentally, the path segment pattern succeeds for incoming requests like `GET /v/1/accounts/[accountId]` that match the path segment pattern exactly.

It would be awesome of these two amazing features played nice together.

**Comment by Pieparker** [comment, marked as answer]
Does [something like this](https://github.com/vercel/vercel/discussions/6730#discussioncomment-1349436) not work for y'all?

- If you're using Next.js, configure [next.config.js](https://nextjs.org/docs/api-reference/next.config.js/rewrites)
- If you're using any other framework, configure [vercel.json](https://vercel.com/docs/cli#project-configuration/rewrites)

It will look like:

```json
{
  "rewrites": [
    { "source": "/:path*", "destination": "/api/:path*" }
  ]
}
```
